src/FromData.purs

```diff
@@ -1,31 +1,142 @@
 import Data.Map as Map
 import Data.Maybe (Maybe(Nothing, Just), maybe)
 import Data.Ratio (Ratio, reduce)
+import Data.Symbol (class IsSymbol, reflectSymbol)
 import Data.Traversable (for, traverse)
 import Data.Tuple (Tuple(Tuple))
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Data.Unfoldable (class Unfoldable)
 import Helpers (bigIntToUInt)
+import Prim.Row as Row
+import Prim.RowList as RL
 import Prim.TypeError (class Fail, Text)
+import Record as Record
+import Type.Proxy (Proxy(Proxy))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Bytes, Constr, List, Map, Integer))
 
+-- | Classes
+
+class FromData :: Type -> Constraint
 class FromData (a :: Type) where
   fromData :: PlutusData -> Maybe a
 
+class FromDataWithIndex :: Type -> Type -> Constraint
+class FromDataWithIndex a ci where
+  fromDataWithIndex :: HasConstrIndex ci => Proxy a -> Proxy ci -> PlutusData -> Maybe a
+
+-- NOTE: Using the 'parser' approach as in https://github.com/purescript-contrib/purescript-argonaut-generic/blob/3ae9622814fd3f3f06fa8e5e58fd58d2ef256b91/src/Data/Argonaut/Decode/Generic.purs
+class FromDataArgs :: Type -> Constraint
+class FromDataArgs a where
+  fromDataArgs :: Array PlutusData -> Maybe { head :: a, tail :: Array PlutusData }
+
+-- | A helper typeclass to implement `ToDataArgs` for records.
+-- Stolen from https://github.com/purescript/purescript-quickcheck/blob/v7.1.0/src/Test/QuickCheck/Arbitrary.purs#L247
+class FromDataArgsRL :: RL.RowList Type -> Row Type -> Constraint
+class FromDataArgsRL list row | list -> row where
+  fromDataArgsRec :: forall rlproxy. rlproxy list -> Array PlutusData -> Maybe { head :: Record row, tail :: Array PlutusData }
+
+-- | FromDataWithIndex instances for Data.Generic.Rep
+-- See https://purescript-simple-json.readthedocs.io/en/latest/generics-rep.html
+
+instance (FromDataWithIndex l a, FromDataWithIndex r a) => FromDataWithIndex (G.Sum l r) a where
+  fromDataWithIndex _ pci pd = G.Inl <$> fromDataWithIndex (Proxy :: Proxy l) pci pd
+    <|> G.Inr <$> fromDataWithIndex (Proxy :: Proxy r) pci pd
+
+instance (IsSymbol n, FromDataArgs arg) => FromDataWithIndex (G.Constructor n arg) a where
+  fromDataWithIndex _ pci (Constr i pdArgs) = do
+    ix <- BigInt.toInt i
+    cn <- resolveConstr pci ix
+    let rn = reflectSymbol (Proxy :: Proxy n)
+    guard $ cn == rn
+    { head: repArgs, tail: pdArgs' } <- fromDataArgs pdArgs
+    guard $ pdArgs' == []
+    pure $ G.Constructor repArgs
+  fromDataWithIndex _ _ _ = Nothing
+
+instance (HasConstrIndex ci, FromDataWithIndex a ci) => FromDataWithIndex (G.Argument a) ci where
+  fromDataWithIndex _ pci pd = G.Argument <$> fromDataWithIndex (Proxy :: Proxy a) pci pd
+
+-- | FromDataArgs instance for Data.Generic.Rep
+
+instance FromDataArgs (G.NoArguments) where
+  fromDataArgs [] = Just { head: G.NoArguments, tail: [] }
+  fromDataArgs _ = Nothing -- TODO: Error "No PlutusData expected but got some"
```

Comment:
I'm a bit confused about the TODO comment: were you intending on adding an actual error here? Pesonally I think it might be better to leave it as `Nothing`. As it is, this instance is total. If we want to attach context/information to failed instances, why not use `Either` instead of `Maybe` for `fromDataArgs`?

Reply:
Exactly, I was using the TODO to mark the places where an error could be caught and reported. I did want to take the liberty to introduce Either in FromData but it definitely could be there.

So if you could help me reach a decision here that would be great; Should we introduce error tracking when parsing Data into Purs ADT (ie. FromData)?

Reply:
I see no reason not to use `Either` instead of `Maybe`. For the moment we can either update `FromData` or just `hush` the error from `fromDataArgs` and update `FromData` later. But I'd definitely be OK with `Either SomeDomainErrorType a` for both classes